web/packages/shared/hooks/useAsync.test.ts

```diff
@@ -212,3 +217,105 @@ test('error and statusText are set when the callback returns a rejected promise'
   // The promise returned from run always succeeds, but any errors are captured as the second arg.
   await expect(runPromise).resolves.toEqual([null, expectedError]);
 });
+
+describe('useDelayedRepeatedAttempt', () => {
+  it('does not update attempt status if it resolves before delay', async () => {
+    let resolve: (symbol: symbol) => void;
+    const { result } = renderHook(() => {
+      const [eagerAttempt, run] = useAsync(() => {
+        // Resolve to a symbol so that successful attempts are not equal to each other.
+        return new Promise<symbol>(res => {
+          resolve = res;
+        });
+      });
+      const delayedAttempt = useDelayedRepeatedAttempt(eagerAttempt, 500);
+
+      return { eagerAttempt, run, delayedAttempt };
+    });
+
+    await act(async () => {
+      const promise = result.current.run();
+      resolve(Symbol());
+      await promise;
+    });
+
+    expect(result.current.delayedAttempt.status).toEqual('success');
+    const oldDelayedAttempt = result.current.delayedAttempt;
+
+    act(() => {
+      // Start promise but do not await it – instead wait for attempt updates. This ensures that we
+      // catch any state updates.
+      result.current.run();
+      resolve(Symbol());
+    });
+
+    // Wait for delayedAttempt to get updated.
+    let nextDelayedAttempt: Attempt<symbol>;
+    await waitFor(
+      () => {
+        // result.current always returns the current result. Capture the attempt that's being
+        // compared within waitFor so that we can check its status outside of the block and be sure
+        // that it's the same attempt.
+        nextDelayedAttempt = result.current.delayedAttempt;
+        expect(nextDelayedAttempt).not.toBe(oldDelayedAttempt);
+      },
+      {
+        onTimeout: error => {
+          error.message =
+            'delayedAttempt did not get updated within timeout. ' +
+            `This might mean that the logic for detecting attempt updates is incorrect.\n${error.message}`;
+          return error;
+        },
+      }
+    );
+
+    // As the promise was resolved before the delay, the attempt status should still be success.
+    expect(nextDelayedAttempt.status).toEqual('success');
```

Comment:
This is all a little bit weird, but with `waitForNextUpdate` being no longer there, I cannot easily test transitions between state updates. So instead I'm depending on the fact that the identity of `result.current.delayedAttempt` will be different between state updates.